text/0255-sensitive-data-redaction.md

```diff
@@ -0,0 +1,276 @@
+- A redactor implements the logic to apply redaction and that owns predefined helpers for redaction in the SDK (URLParams filtering, Zeroing IPs, etc.).
+
+The following limitations apply:
+
+- Only Attributes are treated, although it is possible that sensitive data is contained in other data generated by OpenTelemetry as well (e.g. the span name, or the instrumentation scope name could contain sensitive data)
+
+### Annotate attributes with sensitivity information
+
+As a first building block the OpenTelemetry API needs to provide capability to enrich collected attributes with sensitivity details and with potential hooks to redact those. Those API changes then also can be used to describe sensitivity information in the semantic conventions.
+
+#### API
+
+Every API that sets an attribute consisting of a key and a value, needs to be enhanced by an additional functionality that allows to add details about the potential sensitivity of this data and a hooks how it may be redacted. This can be an additional set of parameters for an existing method or a method that can be called after the attribute has been set, if adding parameters to a method signature would lead to a breaking change.
+
+An additional method for setting the sensitivity information for a span attribute might look like the following:
+
+```
+span.setAttribute("url.query", url.toString(), <SENSITIVITY_DETAILS>);
+```
+
+Or, if the setAttribute method can not be extended:
+
+```
+span.setAttribute("url.query", url.toString());
+span.redactAttribute(“url.query”, <SENSITIVITY_DETAILS>);
+```
+
+The content of `<SENSITIVITY_DETAILS>` may look like the following example:
+
+```
+{
+  <REDACTION_LEVEL>: <REDACTION_RULE>
+}
+```
+
+The key `<REDACTION_LEVEL>` is one of multiple available pre-defined levels of redaction requirements that an end-user may choose, e.g.
+
+- `DEFAULT`: What should be the default redaction applied to this value
+- `STRICT`: What should be a basic redaction that is applied to this value
+- `STRICTER`: What should be an advanced redaction that is applied to this value
+- …
+
+The value `<REDACTION_RULE>` can be one of the following:
+
+- A regular expression (or sed expression?) which when applied turns parts of a string into their redacted version,e.g. `s/([0-9]+\.[0-9]+\.[0-9]+\.)[0-9]+/\10/` applied on an IP address will replace the last octet with 0: `1.2.3.4` becomes `1.2.3.0`
+- A constant that represents pre-defined redaction methods (see below for Redaction helpers), e.g. `REDACT_INSECURE_URL_PARAM_VALUES` will apply what is required for [#971](https://github.com/open-telemetry/semantic-conventions/pull/971)
+- A callback that will call a function on that value and apply advanced redaction
+
+It should be recommended to either use the expression or the constant and only in rare circumstances a callback function should be applied.
```

Comment:
what's the motivation? As a user I'd always prefer a callback to regex for usability, perf, level of control

Reply:
Thanks for calling this out, I didn't add my motivation here. Here was my line of thinking:

This is for **end-users** doing redaction themselves, not for the Semantic Convention. So in their case, the best they can do is reuse some helpers we predefine in the SDK (expressed by constants, eg REDACT_INSECURE_URL_PARAM_VALUES or ANONYMIZE_IP or DROP, ...). I assumed the regex to be second because a regex is much easier to translate them into a helper once again if end-users bring them to the SIG, while a callback can be very convoluted and also have a huge performance impact if not done properly (although I agree the same applies to regular expressions)

Let me rephrase that.

Reply:
I removed that wording for now, it adds no immediate value